datadog_checks_base/datadog_checks/base/checks/win/wmi/sampler.py

```diff
@@ -157,15 +158,60 @@ def _query_sample_loop(self):
             self.logger.info("exception in CoInitialize: %s", e)
             raise
 
+        additional_args = []
+        if self.provider != ProviderArchitecture.DEFAULT:
+            context = Dispatch("WbemScripting.SWbemNamedValueSet")
+            context.Add("__ProviderArchitecture", self.provider)
+            additional_args = [None, "", 128, context]
+
+        locator = Dispatch("WbemScripting.SWbemLocator")
+
         while True:
             self._runSampleEvent.wait()
+            if self._stopping:
+                self.logger.debug("_query_sample_loop stopping")
+                return
+
             self._runSampleEvent.clear()
-            if self.is_raw_perf_class and not self._previous_sample:
-                self._current_sample = self._query()
 
-            self._previous_sample = self._current_sample
-            self._current_sample = self._query()
-            self._sampleCompleteEvent.set()
+            try:
+                # Create a new WMI connection.
+                self.logger.debug(
+                    u"Connecting to WMI server (host=%s, namespace=%s, provider=%s, username=%s).",
+                    self.host,
+                    self.namespace,
+                    self.provider,
+                    self.username,
+                )
+                connection = locator.ConnectServer(
+                    self.host, self.namespace, self.username, self.password, *additional_args
+                )
+
+                if self.is_raw_perf_class and not self._previous_sample:
+                    self._current_sample = self._query(connection)
```

Comment:
I like that _query now receives a connection but I think this would read better if we kept the get_connection method:

```
connection = self.get_connection()
if self.is_raw_perf_class and not self._previous_sample:
                    self._current_sample = self._query(connection)
```

Reply:
That would make it a 1 line method as it would only contain `return locator.ConnectServer(...)`

Bringing the `get_connection()` code into this method allowed me to only create `locator` once rather than on every sample run. I debated whether to just create the connection once as well but decided it was more defensive to do it on each run as I don't know the inners of WMI and whether a connection can be broken.